clang/tools/clang-offload-deps/ClangOffloadDeps.cpp

```diff
@@ -0,0 +1,254 @@
+  if (!ObjectOrErr) {
+    reportError(ObjectOrErr.takeError());
+    return 1;
+  }
+
+  // Then try to find a section in the input binary which contains offload
+  // symbol names and parse section contents.
+  DenseMap<StringRef, SmallDenseSet<StringRef>> Target2Symbols;
+  for (SectionRef Section : ObjectOrErr->getBinary()->sections()) {
+    // Look for the .tgtsym section in the binary.
+    Expected<StringRef> NameOrErr = Section.getName();
+    if (!NameOrErr) {
+      reportError(NameOrErr.takeError());
+      return 1;
+    }
+    if (*NameOrErr != SYMBOLS_SECTION_NAME)
+      continue;
+
+    // This is the section we are looking for, read symbol names from it.
+    Expected<StringRef> DataOrErr = Section.getContents();
+    if (!DataOrErr) {
+      reportError(DataOrErr.takeError());
+      return 1;
+    }
+
+    // Symbol names are prefixed by a target, and prefixed names are separated
+    // by '\0' characters from each other. Find the names matching our list of
+    // offload targets and insert them into the map.
+    for (StringRef Symbol = DataOrErr.get(); !Symbol.empty();) {
+      unsigned Len = strlen(Symbol.data());
+
+      for (const std::string &Target : Targets) {
+        std::string Prefix = Target + ".";
+        if (Symbol.startswith(Prefix))
+          Target2Symbols[Target].insert(
+              Symbol.substr(Prefix.size(), Len - Prefix.size()));
+      }
+
+      Symbol = Symbol.drop_front(Len + 1u);
+    }
+
+    // Binary should not have more than one .tgtsym section.
+    break;
+  }
+
+  LLVMContext Context;
+  Type *Int8PtrTy = Type::getInt8PtrTy(Context);
+
+  // Create bitcode file with the symbol names for each target and write it to
+  // the output file.
+  SmallVector<std::unique_ptr<ToolOutputFile>, 8u> Files;
+  Files.reserve(Outputs.size());
+  for (unsigned I = 0; I < Outputs.size(); ++I) {
+    StringRef FileName = Outputs[I];
+
+    Module Mod{"offload-deps", Context};
+    Mod.setTargetTriple(Triples[I]);
+
+    SmallVector<Constant *, 8u> Used;
+    Used.reserve(Target2Symbols[Targets[I]].size());
+    for (StringRef Symbol : Target2Symbols[Targets[I]])
+      Used.push_back(ConstantExpr::getPointerBitCastOrAddrSpaceCast(
+          Mod.getOrInsertGlobal(Symbol, Int8PtrTy), Int8PtrTy));
+
+    if (!Used.empty()) {
+      ArrayType *ArrayTy = ArrayType::get(Int8PtrTy, Used.size());
+
+      // For SPIRV linking is done on LLVM IR inputs, so we can use special
+      // global variable llvm.used to represent a reference to a symbol. But for
+      // other targets we have to create a real reference since llvm.used may
+      // not be representable in the object file.
+      if (Triple(Triples[I]).isSPIR()) {
+        auto *GV = new GlobalVariable(
+            Mod, ArrayTy, false, GlobalValue::AppendingLinkage,
+            ConstantArray::get(ArrayTy, Used), "llvm.used");
+        GV->setSection("llvm.metadata");
+      } else {
+        auto *GV = new GlobalVariable(
```

Comment:
I think from code maintenance perspective it is better to have single mechanism for SPIRV and non-SPIRV. Will the second work for SPIRV target as well? If yes, can the first one be discarded?

Reply:
Well, LLVM IR just allows to create a reference without adding any additional symbols, I do not see why we should not use this feature. I would rather discard the second part if we had a way to add a reference without adding extra symbols, but looks like this is the only way to do it.

Reply:
AFAIU, both achieve the same result - letting linker know the deps - and only the second can handle all targets. Hence my suggestion. This is a nit anyway.